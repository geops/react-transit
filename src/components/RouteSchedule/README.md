#

This demonstrates the use of RouteSchedule.

```jsx
import React from 'react';
import Dialog from 'react-spatial/components/Dialog';
import BasicMap from 'react-spatial/components/BasicMap';
import Layer from 'react-spatial/Layer';
import TileLayer from 'ol/layer/Tile';
import OSMSource from 'ol/source/OSM';
import TrajservLayer from 'react-transit/layers/TrajservLayer';
import RouteSchedule from 'react-transit/components/RouteSchedule';

class BasicMapExample extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      isRouteScheduleOpen: false,
      lineInfos: null,
    };

    this.layers = [
      new Layer({
        name: 'Layer',
        olLayer: new TileLayer({
          source: new OSMSource(),
        }),
      }),
      new TrajservLayer({
        onClick: trajectory => {
          this.setState({
            isRouteScheduleOpen: true,
            lineInfos: trajectory,
          });
        },
      }),
    ];
  }

  toggleRouteSchedule() {
    const { isRouteScheduleOpen } = this.state;
    this.setState({
      isRouteScheduleOpen: !isRouteScheduleOpen,
    });
  }

  render() {
    const { lineInfos, isRouteScheduleOpen } = this.state;

    return (
      <div className="rt-route-schedule-example">
        <BasicMap center={[951560, 6002550]} zoom={14} layers={this.layers} />
        <Dialog
          className="rt-route-dialog"
          classNameChildren="rt-route-dialog-children"
          onClose={() => this.toggleRouteSchedule()}
          isOpen={isRouteScheduleOpen}
        >
          <RouteSchedule lineInfos={lineInfos} />
        </Dialog>
      </div>
    );
  }
}

<BasicMapExample />;
```
