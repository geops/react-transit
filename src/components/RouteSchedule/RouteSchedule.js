import React from 'react';
import PropTypes from 'prop-types';
import ReactTransitPropTypes from '../../propTypes';
import firstStation from '../../images/RouteSchedule/firstStation.png';
import station from '../../images/RouteSchedule/station.png';
import lastStation from '../../images/RouteSchedule/lastStation.png';
import { bgColors } from '../../config/tracker';

import TrackerLayer from '../../layers/TrackerLayer';

/**
 * Returns a paded number (with leading 0 for integer < 10).
 * @param {Number} number number.
 */
const pad = number => {
  return `${number < 10 ? '0' : ''}${number}`;
};

/**
 * Returns a 'hh:mm' string from a time.
 * @param {Number} t time in milliseconds.
 */
const getTimeString = t => {
  if (t === -1) {
    return '';
  }
  const h = Math.floor(t / 36000000);
  const m = Math.floor((t % 36000000) / 600000);
  return `${pad(h)}:${pad(m)}`;
};

/**
 * Returns a color class to display the delay.
 * @param {Number} time Delay time in milliseconds.
 */
const getDelayString = t => {
  const h = Math.floor(t / 3600000);
  const m = Math.floor((t % 3600000) / 60000);
  const s = Math.floor(((t % 3600000) % 60000) / 1000);

  if (s === 0 && h === 0 && m === 0) {
    return '0';
  }
  if (s === 0 && h === 0) {
    return `${m}m`;
  }
  if (s === 0) {
    return `${h}h${m}m`;
  }
  if (m === 0 && h === 0) {
    return `${s}s`;
  }
  if (h === 0) {
    return `${m}m${s}s`;
  }
  return `${h}h${m}m${s}s`;
};

/**
 * Returns a color class to display the delay.
 * @param {Number} time Delay time in milliseconds.
 */
const getDelayColor = time => {
  const secs = Math.round(((time / 1800 / 2) * 3600) / 1000);
  if (secs >= 3600) {
    return 'dark-red';
  }
  if (secs >= 500) {
    return 'middle-red';
  }
  if (secs >= 300) {
    return 'light-red';
  }
  if (secs >= 180) {
    return 'orange';
  }
  return 'green';
};

/**
 * Returns if the station has already been passed by the vehicule.
 * @param {Object} stop Station information.
 */
const isPassed = (stop, time) => {
  return stop.departureDate * 1000 + stop.departureDelay <= time;
};

/**
 * Returns an image for first, middle or last stations.
 * @param {Number} index index of the station in the list.
 * @param {Number} length Length of the stations list.
 */
const renderStationImg = (index, length) => {
  if (index === 0) {
    return (
      <img
        src={firstStation}
        alt="routeScheduleLine"
        className="rt-route-icon"
      />
    );
  }
  if (index === length - 1) {
    return (
      <img
        src={lastStation}
        alt="routeScheduleLine"
        className="rt-route-icon"
      />
    );
  }
  return (
    <img src={station} alt="routeScheduleLine" className="rt-route-icon" />
  );
};

const propTypes = {
  /**
   * CSS class of the route schedule wrapper.
   */
  className: PropTypes.string,

  /**
   * Trajectory stations informations.
   */
  lineInfos: ReactTransitPropTypes.lineInfos,

  /**
   * Trackerlayer.
   */
  trackerLayer: PropTypes.instanceOf(TrackerLayer).isRequired,

  /**
   * Render Header of the route scheduler.
   */
  header: PropTypes.func,

  /**
   * Render Body of the route scheduler.
   */
  stations: PropTypes.func,

  /**
   * Function triggered on station's click event.
   */
  onStationClick: PropTypes.func,
};

const defaultProps = {
  className: 'rt-route-wrapper',
  lineInfos: null,
  header: null,
  stations: null,
  onStationClick: () => {},
};

const renderRouteIdentifier = (id, longName) => {
  // first part of the id, without leading zeros.
  const routeIdentifier = parseInt(id.split('.')[0], 10);
  if (!longName.includes(routeIdentifier)) {
    return ` (${routeIdentifier})`;
  }
  return null;
};

const renderHeader = lineInfos => (
  <div className="rt-route-header">
    <span
      style={{
        backgroundColor:
          lineInfos.backgroundColor || bgColors[lineInfos.vehiculeType],
        color: lineInfos.color || 'black',
      }}
      className="rt-route-icon"
    >
      {lineInfos.shortName}
    </span>
    <div className="rt-route-title">
      <span className="rt-route-name">{lineInfos.destination}</span>
      <span>
        {lineInfos.longName}
        {renderRouteIdentifier(lineInfos.routeIdentifier, lineInfos.longName)}
      </span>
    </div>
  </div>
);

const renderStations = (lineInfos, onStationClick, trackerLayer) => (
  <div className="rt-route-body">
    {lineInfos.stations.map((stop, idx) => (
      <div
        key={stop.stationId}
        role="button"
        className={`rt-route-station${
          isPassed(stop, trackerLayer.getCurrTime()) ? ' rt-passed' : ''
        }`}
        onClick={e => onStationClick(stop, e)}
        tabIndex={0}
        onKeyPress={e => e.which === 13 && onStationClick(stop, e)}
      >
        <div className="rt-route-delay">
          {typeof stop.arrivalDelay === 'undefined' || idx === 0 ? null : (
            <span
              className={`rt-route-delay-arrival${` ${getDelayColor(
                stop.arrivalDelay,
              )}`}`}
            >
              {`+${getDelayString(stop.arrivalDelay)}`}
            </span>
          )}
          {typeof stop.departureDelay === 'undefined' ||
          idx === lineInfos.stations.length - 1 ? null : (
            <span
              className={`rt-route-delay-arrival${` ${getDelayColor(
                stop.departureDelay,
              )}`}`}
            >
              {`+${getDelayString(stop.departureDelay)}`}
            </span>
          )}
        </div>
        <div className="rt-route-times">
          <span className="rt-route-time-arrival">
            {getTimeString(stop.arrivalTime)}
          </span>
          <span className="rt-route-time-departure">
            {getTimeString(stop.departureTime)}
          </span>
        </div>
        {renderStationImg(idx, lineInfos.stations.length)}
        <div
          className={
            idx === lineInfos.stations.length - 1 && stop.cancelled
              ? 'rt-route-cancelled'
              : null
          }
        >
          {stop.stationName}
        </div>
      </div>
    ))}
  </div>
);

/**
 * RouteSchedule displays information, stops and punctuality about the clicked route.
 */
function RouteSchedule({
  className,
  lineInfos,
  header,
  stations,
  onStationClick,
  trackerLayer,
}) {
  return lineInfos ? (
    <div className={className}>
      {header || renderHeader(lineInfos)}
      {stations || renderStations(lineInfos, onStationClick, trackerLayer)}
    </div>
  ) : null;
}

RouteSchedule.propTypes = propTypes;
RouteSchedule.defaultProps = defaultProps;

export default RouteSchedule;
