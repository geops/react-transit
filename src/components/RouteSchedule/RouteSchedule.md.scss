.rt-route-schedule-example {
  font-family: $font-family;
  position: relative;

  .rt-route-dialog {
    box-shadow: 0 1px 4px rgba(0, 0, 0, 0.5);
    background: white;
    position: absolute;
    margin: auto;
    left: 30px;
    top: 30px;
    height: 80%;
    max-height: 80%;

    .tm-dialog-header {
      position: absolute;
      cursor: pointer;
      top: 10px;
      right: 10px;
      padding: 5px;
    }

    .rt-route-dialog-children {
      height: 100%;

      .rt-route-wrapper {
        height: 100%;

        .rt-route-body {
          max-height: calc(100% - 80px);
          height: calc(100% - 80px);
        }
      }
    }
  }
}
