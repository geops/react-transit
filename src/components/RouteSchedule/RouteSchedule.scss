.rt-route-wrapper {
  background-color: white;
  width: 350px;
  overflow: hidden;

  .rt-route-header {
    display: flex;
    align-items: center;
    padding: 15px 10px 0 10px;

    .rt-route-title {
      display: flex;
      flex-direction: column;

      .rt-route-name {
        padding-bottom: 8px;
        font-weight: bold;
      }
    }

    .rt-route-icon {
      -webkit-border-radius: 20px;
      -moz-border-radius: 20px;
      border-radius: 20px;
      min-width: 20px;
      height: 20px;
      line-height: 20px;
      border: solid black 2px;
      padding: 5px;
      display: block;
      float: left;
      margin: 15px;
      margin-top: 10px;
      text-align: center;
      font-size: 14px;
      font-weight: bold;
    }
  }

  .rt-route-body {
    font-size: 14px;
    max-height: 650px;
    overflow: auto;
    padding: 10px 20px;

    .rt-route-station {
      display: flex;
      align-items: center;
      cursor: pointer;
      border-radius: 4px;

      &:hover {
        color: white;
        background-color: $primary-blue;
      }

      .rt-route-times,
      .rt-route-delay {
        display: flex;
        flex-direction: column;
        width: 40px;
        min-width: 40px;
        padding: 0 3px;

        .green {
          color: rgb(0, 160, 12);
        }

        .orange {
          color: rgb(247, 191, 0);
        }

        .light-red {
          color: rgb(255, 74, 0);
        }

        .middle-red {
          color: rgb(232, 0, 0);
        }

        .dark-red {
          color: rgb(237, 0, 76);
        }
      }

      .rt-route-icon {
        padding: 0 15px;
      }
    }

    .rt-route-station:first-child {
      .rt-route-time-arrival {
        display: none;
      }
    }

    .rt-route-station:last-child {
      .rt-route-time-departure {
        display: none;
      }
    }
  }
}
