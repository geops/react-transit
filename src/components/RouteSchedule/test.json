{
  "ID": 9976006, // id
  "VehiculeType": 0, // t
  "WheelchairAccessible": 0, // wa
  "BicyclesAllowed": 0, // ba
  "BackgroundColor": "ff8a00", // c
  "WritingColor": "", // tc
  "Destination": "Zürich, Laubegg", // hs
  "RealTime": 0, // rt
  "ShortName": "5", // sn
  "LongName": "T 5", // ln
  "FeedsID": 580, // fid
  "OperatingInformations": { // tt
    "OperatingPeriod": 31, // t
    "NotOperatingDays": [ // n
      {
        "Month": 12, // m
        "Day": 25, // d
        "Year": 118 // y
      }
    ],
    "AdditionalOperatingDays": [ // p
      {
        "Month": 12, // m
        "Day": 15, // d
        "Year": 119 // y
      }
    ]
  },
  "Stations": [
    {
      "Coordinates": [8.68696,47.42572], // p
      "ArrivalTime": 606600000, // at
      "DepartureTime": 606600000, // dt
      "ArrivalDate": 1565621460, // ap
      "DepartureDate": 1565621460, // dp
      "ArrivalDelay": 60000, // ad
      "DepartureDelay": 60000, // dd
      "Skipped": 0, // c
      "StationID": "8503305:6", // sid
      "NoDropOff": 0, // dot, if 1 no stop
      "NoPickup": 0, // put if 1 can't enter
      "Name": "Effretikon", // n
      "WheelchairAccessible": 0 // wa
    }
  ],
  "Operator": { // a
    "Name": "VBZ    F (Verkehrsbetriebe Zürich INFO+)", // n
    "Url": "http://www.sbb.ch", // u
    "l": "",
    "p": "",
    "f": "",
    "TimeZone": "Europe/Berlin" // tz
  },
  "f": {
    "Name": "SBB", // n
    "Url": "http://www.sbb.ch", // u
    "Language": "de", // l
    "s": "20181209",
    "e": "20191214",
    "v": "Fahrplan 2019"
  }
}