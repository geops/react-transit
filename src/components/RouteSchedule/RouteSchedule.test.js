import React from 'react';
import renderer from 'react-test-renderer';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import RouteSchedule from '.';

configure({ adapter: new Adapter() });

const lineInfos = {
  backgroundColor: 'ff8a00',
  destination: 'Station name',
  id: 9959310,
  longName: 'T 3',
  shortName: '3',
  stations: [
    {
      stationId: '1',
      stationName: 'first stop',
      coordinates: [8.51772, 47.3586],
      arrivalDelay: 60000,
      arrivalTime: 603000000,
      departureDelay: 60000,
      departureTime: 603000000,
    },
    {
      stationId: '2',
      stationName: 'second stop',
      coordinates: [8.54119, 47.36646],
      arrivalDelay: 120000,
      arrivalTime: 609000000,
      departureDelay: 120000,
      departureTime: 609000000,
    },
  ],
  vehiculeType: 0,
};

describe('RouteSchedule', () => {
  test('matches snapshots.', () => {
    const component = renderer.create(<RouteSchedule lineInfos={lineInfos} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  // to test: on station click
  // to test: time formating
  // to test: delay formating
  // to test: delay color
});
