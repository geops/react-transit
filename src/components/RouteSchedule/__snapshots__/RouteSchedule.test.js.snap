// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RouteSchedule matches snapshots. 1`] = `
<div
  className="rt-route-wrapper"
>
  <div
    className="rt-route-header"
  >
    <span
      className="rt-route-icon"
      style={
        Object {
          "backgroundColor": "ff8a00",
          "color": "black",
        }
      }
    >
      3
    </span>
    <div
      className="rt-route-title"
    >
      <span
        className="rt-route-name"
      >
        Station name
      </span>
      <span>
        T 3
      </span>
    </div>
  </div>
  <div
    className="rt-route-body"
  >
    <div
      className="rt-route-station"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="rt-route-delay"
      >
        <span
          className="rt-route-delay-arrival green"
        >
          +01m
        </span>
        <span
          className="rt-route-delay-arrival green"
        >
          +01m
        </span>
      </div>
      <div
        className="rt-route-times"
      >
        <span
          className="rt-route-time-arrival"
        >
          16:45
        </span>
        <span
          className="rt-route-time-departure"
        >
          16:45
        </span>
      </div>
      <img
        alt="routeScheduleLine"
        className="rt-route-icon"
        src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAyCAYAAABcfPsmAAAABmJLR0QA/wD/AP+gvaeTAAAACXBIWXMAAAsTAAALEwEAmpwYAAAAB3RJTUUH3gYeCBQzRAvlCAAAABl0RVh0Q29tbWVudABDcmVhdGVkIHdpdGggR0lNUFeBDhcAAAD2SURBVEjH7ZPNEYIwEEafPwXQgSnBBhydoRDog4N44GAVQiMODBXQAcPVCx3owUUzmjAyenO/S0J280h2v4BKpVKp/lQzb2STGcAAW2u1AlrqpJ0GvMP2AtxZkRJogYMPuhiBxQK0ZYA1ELAKG7pz/7p96fjJTmDIaSoZh+sbiVdA/rp57gBG1rwAUuokBVL5duWNAu2a5Y9a3cfckzcK/EouYGnNY2nS0KzYkzfalMK6TgQYNlnr8GTxKbCUWg22MY6c3HfCdx92555V2ACBeM4F8xrb15QLcPLEjtOf3vPVXN/W6mQ2tcs/t40CFahABSrwB7oBXXk7PEfz+n8AAAAASUVORK5CYII="
      />
      <div>
        first stop
      </div>
    </div>
    <div
      className="rt-route-station"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="rt-route-delay"
      >
        <span
          className="rt-route-delay-arrival green"
        >
          +02m
        </span>
        <span
          className="rt-route-delay-arrival green"
        >
          +02m
        </span>
      </div>
      <div
        className="rt-route-times"
      >
        <span
          className="rt-route-time-arrival"
        >
          16:55
        </span>
        <span
          className="rt-route-time-departure"
        >
          16:55
        </span>
      </div>
      <img
        alt="routeScheduleLine"
        className="rt-route-icon"
        src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAyCAYAAABcfPsmAAAABmJLR0QA/wD/AP+gvaeTAAAACXBIWXMAAAsTAAALEwEAmpwYAAAAB3RJTUUH3gYeCBMwkkMidQAAABl0RVh0Q29tbWVudABDcmVhdGVkIHdpdGggR0lNUFeBDhcAAAEeSURBVEjH7ZSxbcMwEEWflAzgJrVZpswCRgR4ECtTpGARpfAe8iKGDNYpXKUVPIGa1E5h2iBkHkMCApLA/IAgSvfv607/yIIQFuvj1Tuji1BKycSYXLAItAtwFHOMTqjwJDZRy7FiAu/vm5IF/4WgsANieWUKOSZeJif98LEbdDl0Hj4Cn57IM/CB0V8p56ECXoVPvQAPUh13gtgbUAs5T8CM+XLPYTuMg/eehMoR64GdvSvbrrLxHdDGtLxy1hugwegGaOyzjxcUrJx1i9G9Heh+VFH1a3PYOevamnQ2qxZ4QVM2TjsrQLFYu6a4vCjBzv6r2oooD6eVKryew8N2YL7cAzNgGAl29nq/mJWw9ZSnzdNMCmIZGRkZGTeNb4FoRhP3GY83AAAAAElFTkSuQmCC"
      />
      <div>
        second stop
      </div>
    </div>
  </div>
</div>
`;
